src/quicklist.c

```diff
@@ -1092,6 +1004,70 @@ void quicklistInsertAfter(quicklistIter *iter, quicklistEntry *entry,
     _quicklistInsert(iter, entry, value, sz, 1);
 }
 
+/* Replace quicklist entry by 'data' with length 'sz'. */
+void quicklistReplaceEntry(quicklistIter *iter, quicklistEntry *entry,
+                           void *data, size_t sz)
+{
+    quicklist *quicklist = iter->quicklist;
+    quicklistNode *node = entry->node;
+
+    if (likely(!QL_NODE_IS_PLAIN(node) && !isLargeElement(sz) && 
+               (node->count == 1 || !quicklistNodeExceedsLimit(quicklist->fill, 
+                node->sz - entry->sz + sz + SIZE_ESTIMATE_OVERHEAD, node->count)))) {
+        node->entry = lpReplace(node->entry, &entry->zi, data, sz);
+        quicklistNodeUpdateSz(node);
+        /* quicklistNext() and quicklistGetIteratorEntryAtIdx() provide an uncompressed node */
+        quicklistCompress(quicklist, node);
+    } else if (node->count == 1) {
+        zfree(node->entry);
+        if (isLargeElement(sz)) {
+            node->entry = zmalloc(sz);
+            memcpy(node->entry, data, sz);
+            node->sz = sz;
+            node->container = QUICKLIST_NODE_CONTAINER_PLAIN;
+        } else {
+            node->entry = lpPrepend(lpNew(0), data, sz);
+            quicklistNodeUpdateSz(node);
+            node->container = QUICKLIST_NODE_CONTAINER_PACKED;
+        }
+        quicklistCompress(quicklist, node);
+    } else {
+        /* node->count > 1, node will not be removed */
+        quicklistDelEntry(iter, entry);
+        if (quicklistNext(iter, entry)) {
+            _quicklistInsert(iter, entry, data, sz, iter->direction == AL_START_HEAD);
+        } else {
+            int direction = iter->direction == AL_START_HEAD 
+                            ? AL_START_TAIL : AL_START_HEAD;
+            quicklistIter *rev_iter = quicklistGetIterator(quicklist, direction);
+            quicklistNext(rev_iter, entry);
+            _quicklistInsert(rev_iter, entry, data, sz, direction == AL_START_HEAD);
+            quicklistReleaseIterator(rev_iter);
```

Comment:
I wouldn't say I like this way that nested use of iterators.
Can you explain why we have to do as this.

Reply:
We can't insert first now, because entry->node have a chance be freed due to merge.

Reply:
In that case I'd prefer an implementation like _quicklistInsert to reimplement the replacement, rather than nested iterators.
But that would introduce a lot of duplicate code and more complexity.